interop-reactiveStreams/shared/src/main/scala/scalaz/zio/interop/reactiveStreams/QueueSubscriber.scala

```diff
@@ -1,56 +1,77 @@
-        case _                                     => UIO.unit
-      }
+import scalaz.zio.stream.ZStream
+import scalaz.zio.stream.ZStream.Fold
+import scalaz.zio.{ IO, Queue, Runtime, Task, UIO, ZIO }
+
+private[reactiveStreams] object QueueSubscriber {
+  def make[A](capacity: Int): ZIO[Any, Nothing, (Subscriber[A], ZStream[Any, Throwable, A])] =
+    for {
+      runtime <- UIO.runtime[Any]
+      q       <- Queue.bounded[A](capacity)
+      qs      = new QueueSubscriber[A](runtime, q)
+    } yield (qs, qs)
+}
+
+private class QueueSubscriber[A](runtime: Runtime[_], q: Queue[A])
+    extends Subscriber[A]
+    with ZStream[Any, Throwable, A] {
+
+  private val capacity: Long = q.capacity.toLong
+
+  private var completed: Boolean                 = false
+  private var failed: Option[Throwable]          = None
+  private var subscription: Option[Subscription] = None
+
+  override def fold[R1 <: Any, E1 >: Throwable, A1 >: A, S]: Fold[R1, E1, A1, S] =
+    IO.succeedLazy { (s, cont, f) =>
+      def loop(s: S, demand: Long): ZIO[R1, E1, S] =
+        if (!cont(s)) subscription.fold(UIO.unit)(s => UIO(s.cancel())) *> q.shutdown.const(s)
+        else
+          q.size.flatMap { n =>
+            val empty = n <= 0
+            if (empty && completed) q.shutdown.const(s)
+            else if (empty && failed.isDefined) q.shutdown *> Task.fail(failed.get)
+            else if (empty && (demand < q.capacity))
+              subscription.fold(UIO.unit)(s => UIO(s.request(capacity - demand))) *> loop(s, capacity)
+            else q.take.flatMap(f(s, _)).flatMap(loop(_, demand - 1))
+          } <> failed.fold[Task[S]](UIO.succeed(s))(Task.fail)
+      loop(s, capacity)
     }
 
-  override def onSubscribe(subscription: Subscription): Unit = {
-    if (subscription == null) throw new NullPointerException("s was null in onSubscribe")
-    if (subscriptionOpt.isDefined) {
-      subscription.cancel()
-    } else {
-      subscriptionOpt = Some(subscription)
-      runtime.unsafeRunAsync_(q.awaitShutdown *> UIO(subscription.cancel()))
-      subscription.request(q.capacity.toLong)
+  override def onSubscribe(s: Subscription): Unit = {
+    if (s == null) throw new NullPointerException("s was null in onSubscribe")
+    synchronized {
+      subscription.fold {
+        subscription = Some(s)
+        s.request(capacity)
+      }(_ => s.cancel())
     }
   }
 
   override def onNext(t: A): Unit = {
     if (t == null) throw new NullPointerException("t was null in onNext")
-    runtime.unsafeRunSync(q.offer(Value(t)))
-    signalledDemand -= 1
+    runtime.unsafeRunSync(q.offer(t))
+    ()
   }
 
   override def onError(e: Throwable): Unit = {
     if (e == null) throw new NullPointerException("t was null in onError")
-    signalledDemand = 0
-    subscriptionOpt = None
-    runtime.unsafeRunSync(q.offer(Fail(e)).void)
-    ()
+    if (failed.isEmpty) {
+      failed = Some(e)
+      shutdownQueueIfEmpty()
```

Comment:
I think it's OK to shutdown the queue immediately rather than waiting for it to drain. Whoever is signalling the error can always handle it on their side and signal a completion instead.